### PR TITLE
autoencoder ML implementation using Yelp data subset

### PR DESCRIPTION
**Pull request by sganes21**
*** Pushing from personal sganes branch to Dev (avoiding pushes to Main until integration strategy discussion)

Notes:
Autoencoder approach to subset of Yelp dataset for training and testing. (Can be expanded to entire dataset)  

Files added:
**requirements.txt** 
- Using pytorch libraries for encoder implementation. running using standard cpu hw. Will need to be modified for CUDA GPU support.

**Scripts:**

data_preprocessor.py: Runs the entire implementation workflow:
- Takes in data (Yelp dataset stored locally) as a pandas dataframe
- cleans data and performs simple preprocessing 
- splits into training/testing data
- generates model from training data subset
- applies autoencoder solution to training data subset
- Trains the data using autoencoder
- Generates embedding from test set
- Generates recommendation based on embedding 

**Functions:**

dataloader.py: Function to load locally stored .json files
- Parameters (directory, filename)
- Returns: dataframe

autoencoder.py: Function using pytorch that returns denoising autoencoder model
- Parameters (number of features, encoding dimension)
- Returns: autoencoder_model 
Denoising applied in order to randomly mask subset of values during training, to prevent overfitting of model

autoencoder_training.py: 
- Parameters (autoencoder_model, X_train, epochs, batch_size )
- Returns: trained model
Trains the autoencoder using the specified model and training set. parameters are set to default in the main data_preprocessor.py script.
Employs an optimizer for efficiency

autoencoder_testing.py
- Parameters (trained_model, X_test, dataframe)
- Returns: embeddings
Tests the autoencoder trained model

recommender.py
- Parameters(embeddings, dataframe, query_idx, number of recommendations)
- Returns: recommendations
Note: number of recommendations hardcoded to 5
Note:  query_idx hardcoded to 0

### CODE REVIEW

**Comment by JesseFWarren** [comment]
Thanks for getting this set up Santosh! this is a solid foundation for the classical ML pipeline! 
I like that you have a clear structure with modular scripts for training, testing, and recommendation generation.
For integrating this to the Render frontend it would be great to return results as a list of JSONs like:
{"name": "...", "category": "...", "rating": 4.3, "price": 2, "address": "..."}
We could add a formatter function at the end of the pipeline to prep the frontend output.

I'll start working on the next steps on the backend connector to integrate this with our frontend!

**Comment by dayeon3117** [comment]
This looks awesome Santosh! The structure is super clean and easy to follow. I really like how modular each script is. It’ll make integration and testing way easier. The way data_preprocessor.py pulls everything together is really clear too.

Your notes in the PR were helpful and made it easy to understand how the pipeline works from start to finish. The defaults make sense and I like that it’s flexible if we want to change things later. This sets us up really well to scale the classical model!

**Comment by rajivsraman** [comment]
Thank you for handling the autoencoder pipeline Santosh! Based on the Render application, it seems like your approach works quite well. Also, it was interesting to read up on autoencoders and how they can reconstruct original data by encoding the input in a latent space. Seems like a very cool approach for learning potentially missing interactions in the data!
